### ISSUE TITLE
Questions about int8 inference procedure

### ISSUE THREAD

**Issue by DaeHwanGi** [NVIDIA/TensorRT#1271]
## Description
I'm trying to serve my model with TensorRT. I followed this document https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/userguide.html#export-to-onnx. Here, I realize Q/DQ layer in onnx is not fully supported yet (#994)

1. After QAT I could get amax at activation and weight, to perform int8 inference, set input_tensor dynamic_range in https://github.com/NVIDIA/TensorRT/blob/release/7.2/demo/BERT/builder.py, but how can I set weigth dynamic_range per-channel quantization
   1. Does this method support automatic fusion?
   2. Is there any way to modify the parsed onnx network? (instead of copying the weight)
2. how can i check cetain layer is performed with int8

Additionally when i upate TensorRT to 8.0.0, I saw the log below. Any clue to solve this error or where I can look into? Any help would be much appreciated.

~~~
[TensorRT] ERROR: 2: [optimizer.cpp::isPolymorphic::1018] Error Code 2: Internal Error (Assertion asCopyingLeafNode() ? candidateChoices.empty() : !candidateChoices.empty() failed.)
~~~

## Environment

**TensorRT Version**: 7.2.3.4
**NVIDIA GPU**:  V100
**NVIDIA Driver Version**: 450.80.02
**CUDA Version**: 11.1
**CUDNN Version**: 8.1.0
**Operating System**: ubuntu 18.04
**Python Version (if applicable)**: 3.8.8
**PyTorch Version (if applicable)**: 1.8.1

**Comment by ttyio** [comment]
Hello @DaeHwanGi ,

> how can I set weigth dynamic_range per-channel quantization
We support per tensor for activation and per channel for weights, and because weights calibration is TRT built-in functionality for PTQ, so we don't provide API to change it. You only need to set the scale for activations.

> Does this method support automatic fusion?
Yes.

> Is there any way to modify the parsed onnx network? (instead of copying the weight)
Are you looking for `get_layer` in `INetworkDefinition` ?  https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Network.html#

> how can i check cetain layer is performed with int8
Build the engine with verbose log level

> Any clue to solve this error or where I can look into?
Could you share us with repro? thanks

**Comment by ttyio** [comment]
@DaeHwanGi 
We support D/DQ, in 8.0 release, could you also take a try with 8.0EA? thanks

**Comment by ttyio** [comment]
sorry @DaeHwanGi  I thought this is error from 7.2, this is an internal error message. could you share us the onnx file to debug? thanks

**Comment by DaeHwanGi** [comment]
@ttyio 
I sent my onnx file to your email.

**Comment by DaeHwanGi** [comment]
@ttyio I also keep trying to manually convert to int8 with TensorRT 7.2.3.4.
I didn't use the qkvToContext plugin because of the below constraint.
I use a 48 sequence length as an input.
https://github.com/NVIDIA/TensorRT/blob/4c99d07241d9bc1be1d3727be5f3fa3ffd5e755a/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L187-L192

In the documentation(https://docs.nvidia.com/deeplearning/tensorrt/archives/tensorrt-723/developer-guide/index.html#set_layer_mp_python). I checked the below code need for int8 inference.
```
builder.strict_type_constraints = true
...
layer.precision = trt.int8
layer.set_output_type(out_tensor_index, trt.int8)

dr_input = weights_dict[layer.name+'_input_quantizer._amax']
input_tensor = layer.get_input(0)
input_tensor.dynamic_range = (-dr_input, dr_input)

dr_input = weights_dict[next_layer.name+'_input_quantizer._amax']
output_tensor = layer.get_output(0)
output_tensor.dynamic_range = (-dr_input, dr_input)
``` 

When I tried to convert manually, I saw the log below and it inferred with float precision.
>[TensorRT] WARNING: No implementation of layer in_proj obeys the requested constraints in strict mode. No conforming implementation was found i.e. requested layer computation precision and output precision types are ignored, using the fastest implementation.

I expect that additional conditions are necessary. Am I need to set setAllowedFormats for all inputs?

**Comment by ttyio** [comment]
Hello @DaeHwanGi , thanks for sharing the model, the fix for the importing from ONNX model will be available in the 8.0GA.

For the pack size check, we have fused the multi head attention as single kernel when len = 128, 384 for fixed-seq-len version, and for your len = 48, there is no fused kernel implementation, and the unfused version has no INT8 support.

For all layers that has boolean or loop, the backend has no support for INT8 in 7.2. And if there is no precision issue and you have to use partial quantization, you can disable the `strict_type_constraints`, and allow TRT to select the fastest implementation.

**Comment by ttyio** [comment]
> @ttyio
> I sent my onnx file to your email.

Hello @DaeHwanGi , this is fixed in 8.0GA, thanks

**Comment by ttyio** [comment]
Closing since no activities for more than 3 weeks, please reopen if you still have question. Thanks!
